skyportal/handlers/api/user_obj_list.py

```diff
@@ -0,0 +1,356 @@
+        parameters:
+        - in: path
+          name: listing_id
+          required: true
+          schema:
+            type: integer
+        requestBody:
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  user_ud:
+                    type: integer
+                    required: true
+                  obj_id:
+                    type: string
+                    required: true
+                  list_name:
+                    type: string
+                    required: true
+                    description: |
+                        Listing name for this item, e.g., "favorites".
+                        Multiple objects can be saved by the same user to different
+                        lists, where the list names are user-defined.
+                        List name must be a non-empty string starting with an
+                        alphanumeric character or underscore.
+                        (it must match the regex: /^\\w+/)
+
+        responses:
+          200:
+            content:
+              application/json:
+                schema: Success
+
+        """
+
+        listing = Listing.query.get(int(listing_id))
+
+        if listing is None:
+            return self.error("Listing does not exist.")
+
+        # verify that poster has write access to user_id's lists
+        if (
+            not self.associated_user_object.id == listing.user_id
+            and "System admin" not in self.associated_user_object.permissions
+        ):
+            return self.error('Insufficient permissions to access this listing. ')
+
+        # get the data from the request body
+        data = self.get_json()
+
+        schema = Listing.__schema__()
+        try:
+            schema.load(data)
+        except ValidationError as e:
+            return self.error(f'Invalid/missing parameters: {e.normalized_messages()}')
+
+        user_id = data.get('user_id')
+
+        if User.query.get(user_id) is None:  # verify that user exists
+            return self.error(f'User "{user_id}" does not exist!')
+
+        user_id = int(user_id)
+
+        # verify that poster has write access to the new user_id
+        if (
+            not self.associated_user_object.id == user_id
+            and "System admin" not in self.associated_user_object.permissions
+        ):
+            return self.error(
+                f'Insufficient permissions to change listings for user ID: {user_id}. '
+            )
+
+        obj_id = data.get('obj_id')
+        if Obj.query.get(obj_id) is None:  # verify that object exists!
+            return self.error(f'Object "{obj_id}" does not exist!')
+
+        list_name = data.get('list_name')
+        if not re.search(r'^\w+', list_name):
+            return self.error(
+                "Input `list_name` must begin with alphanumeric/underscore"
+            )
```

Comment:
this rule should only be written in one place

Reply:
You mean add the validation in the marshmallow schema? I'm not sure how to do that...